yarpctest/recorder/recorder.go

```diff
@@ -0,0 +1,324 @@
+// environment:
+//
+// 	dispatcher, err := yarpc.NewDispatcher(yarpc.Config{name: ...},
+// 		xyarpc.Filter(recorder.NewRecorder(t)))
+//
+// The recorded messages will be stored in
+// `./testdata/recordings/*.yaml`.
+func NewRecorder(logger Logger) (recorder *Recorder) {
+	cwd, err := os.Getwd()
+	if err != nil {
+		logger.Fatal(err)
+	}
+	mode, err := ModeFromString(*recorderFlag)
+	if err != nil {
+		logger.Fatal(err)
+	}
+	recorder = &Recorder{
+		mode:       invalidMode,
+		logger:     logger,
+		recordsDir: filepath.Join(cwd, recorderDir),
+	}
+	recorder.SetMode(mode)
+	return recorder
+}
+
+// SetMode let you choose enable the different replay and recording modes,
+// overriding the --recorder flag.
+func (r *Recorder) SetMode(newMode Mode) {
+	if r.mode == newMode {
+		return
+	}
+	r.mode = newMode
+	r.logger.Logf("recorder %s to %v", r.mode.toHumanString(), r.recordsDir)
+}
+
+func sanitizeFilename(s string) (r string) {
+	const allowedRunes = `_-.`
+	return strings.Map(func(rv rune) rune {
+		if unicode.IsLetter(rv) || unicode.IsNumber(rv) {
+			return rv
+		}
+		if strings.ContainsRune(allowedRunes, rv) {
+			return rv
+		}
+		return '_'
+	}, s)
+}
+
+func (r *Recorder) hashRequest(request *transport.Request, body []byte) string {
+	log := r.logger
+	hash := fnv.New64a()
+
+	ha := func(b string) {
+		_, err := hash.Write([]byte(b))
+		if err != nil {
+			log.Fatal(err)
+		}
+		_, err = hash.Write([]byte("."))
+		if err != nil {
+			log.Fatal(err)
+		}
+	}
+
+	ha(request.Caller)
+	ha(request.Service)
+	ha(string(request.Encoding))
+	ha(request.Procedure)
+
+	headersMap := request.Headers.Items()
+	orderedHeadersKeys := make([]string, 0, len(headersMap))
+	for k := range headersMap {
+		orderedHeadersKeys = append(orderedHeadersKeys, k)
+	}
+	sort.Strings(orderedHeadersKeys)
+	for _, k := range orderedHeadersKeys {
+		ha(k)
+		ha(headersMap[k])
+	}
+
+	ha(request.ShardKey)
+	ha(request.RoutingKey)
+	ha(request.RoutingDelegate)
+
+	_, err := hash.Write(body)
+	if err != nil {
+		log.Fatal(err)
+	}
+	return fmt.Sprintf("%x", hash.Sum64())
+}
+
+func (r *Recorder) makeFilePath(request *transport.Request, hash string) string {
+	s := fmt.Sprintf("%s.%s.%s.yaml", request.Service, request.Procedure, hash)
+	return filepath.Join(r.recordsDir, sanitizeFilename(s))
+}
+
+// Call implements the yarpc transport filter interface
+func (r *Recorder) Call(
+	ctx context.Context,
+	request *transport.Request,
+	out transport.Outbound,
+) (*transport.Response, error) {
+	log := r.logger
+	requestBody, err := ioutil.ReadAll(request.Body)
+	request.Body = bytes.NewReader(requestBody)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	requestHash := r.hashRequest(request, requestBody)
+	filepath := r.makeFilePath(request, requestHash)
+
+	switch r.mode {
+	case Replay:
+		return r.loadRecord(filepath)
+	case Append:
+		response, err := r.loadRecord(filepath)
+		if !os.IsNotExist(err) {
+			return response, err
```

Comment:
- loadRecord does't wrap it, by design, because of the next point.
 - using stat is a potential race condition.
 - os.IsNotExist(err) is effectively testing if the file did not exist at the time of the file open.